modules/ssl/ssl_engine_pphrase.c

```diff
@@ -600,3 +610,330 @@ int ssl_pphrase_Handle_CB(char *buf, int bufsize, int verify, void *srv)
+                            "Init: Failed to create pass phrase pipe '%s'",
+                            sc->server->pphrase_dialog_path);
+                    //TODO fix this error
+                    PEMerr(PEM_F_PEM_DEF_CALLBACK,
+                            PEM_R_PROBLEMS_GETTING_PASSWORD);
+                    return 0;
+                }
+            }
+            ap_log_error(APLOG_MARK, APLOG_INFO, 0, ppcb_arg->s, APLOGNO(01967)
+                    "Init: Requesting pass phrase via piped dialog");
+        }
+        else { /* sc->server->pphrase_dialog_type == SSL_PPTYPE_BUILTIN */
+#ifdef WIN32
+            //TODO Fix this error
+            PEMerr(PEM_F_PEM_DEF_CALLBACK, PEM_R_PROBLEMS_GETTING_PASSWORD);
+            return 0;
+#else
+            /*
+             * stderr has already been redirected to the error_log.
+             * rather than attempting to temporarily rehook it to the terminal,
+             * we print the prompt to stdout before EVP_read_pw_string turns
+             * off tty echo
+             */
+            apr_file_open_stdout(&writetty, ppcb_arg->p);
+
+            ap_log_error(APLOG_MARK, APLOG_INFO, 0, ppcb_arg->s, APLOGNO(01968)
+                    "Init: Requesting pass phrase via builtin terminal "
+                    "dialog");
+#endif
+        }
+
+        /*
+         * The first time display a header to inform the user about what
+         * program he actually speaks to, which module is responsible for
+         * this terminal dialog and why to the hell he has to enter
+         * something...
+         */
+        if (ppcb_arg->nPassPhraseDialog == 1) {
+            apr_file_printf(writetty, "%s mod_ssl (Pass Phrase Dialog)\n",
+                    AP_SERVER_BASEVERSION);
+            apr_file_printf(writetty, "Some of your private key files are encrypted for security reasons.\n");
+            apr_file_printf(writetty, "In order to read them you have to provide the pass phrases.\n");
+        }
+        if (ppcb_arg->bPassPhraseDialogOnce) {
+            ppcb_arg->bPassPhraseDialogOnce = FALSE;
+            apr_file_printf(writetty, "\n");
+            apr_file_printf(writetty, "Private key %s (%s)\n",
+                    ppcb_arg->key_id, ppcb_arg->pkey_file);
+        }
+    }
+    
+    return 1;
+}
+
+int passphrase_ui_read(UI *ui, UI_STRING *uis)
+{
+    pphrase_cb_arg_t *ppcb_arg;
+    SSLSrvConfigRec *sc;
+    const char *prompt;
+    int i;
+    int bufsize;
+    int len;
+    char *buf;
+
+    ppcb_arg = (pphrase_cb_arg_t *) UI_get0_user_data(ui);
+
+    if (ppcb_arg == NULL) {
+        //TODO LOG Properly
+        return 0;
+    }
+
+    sc = mySrvConfig(ppcb_arg->s);
+
+    prompt = UI_get0_output_string(uis);
+    if (prompt == NULL) {
+        prompt = "Enter pass phrase:";
+    }
+
+    /*
+     * Get the maximum expected size and allocate the buffer
+     */
+    bufsize = UI_get_result_maxsize(uis);
+    buf = (char *)OPENSSL_zalloc(bufsize);
```

Comment:
Shouldn't that use apr's allocation functions?

Reply:
I was not sure about this, because this buffer is used inside the engine. But yes, since it is a buffer provided by the application, it should be managed by the application. I will change it to use the apr allocation functions.